Engine/KDtree.h

```diff
@@ -0,0 +1,503 @@
+//inherits primitive
+
+//First build a ez tree
+//accelerate by turn Triangle to pointer
+//primitive bound make first
+//traversal use bit shift
+namespace ph{
+using namespace std;
+
+class Vec3 {
+	public:
+		float x;
+		float y;
+		float z;
+		//friend is needed, without friend there will be an implicit parameters this which cause compile error
+		friend Vec3 operator-(Vec3 lhs, const Vec3& rhs){
+			Vec3 temp;
+			temp.x = lhs.x - rhs.x;
+			temp.y = lhs.y - rhs.y;
+			temp.z = lhs.z - rhs.z;
+			return temp;
+		}
+
+		Vec3(){
+
+		}
+
+		Vec3(float in_x, float in_y, float in_z){
+			x = in_x;
+			y = in_y;
+			z = in_z;
+		}
+
+		void setVec3(float in_x, float in_y, float in_z){
+			x = in_x;
+			y = in_y;
+			z = in_z;
+		}
+};
+
+float Three_Vec3_delta(const Vec3 A, const Vec3 B, const Vec3 C){
+	return A.x * B.y * C.z + B.x * C.y * A.z + C.x * A.y * B.z + C.x * A.y * B.z
+			- A.z * B.y * C.z - B.z * C.y * A.x - C.z * A.y * B.x;
+}
+
+
+class AABB{
+	private:
+		Vec3 VertexMin;
+		Vec3 VertexMax;
+		bool _is_set = 0;
+		bool min_set = 0;
+		bool max_set = 0;
+	public:
+		AABB(){
+			_is_set = 0;
+			min_set = 0;
+			max_set = 0;
+		}
+		bool is_set(){
+			return _is_set;
+		}
+		void setVertexMin(float x, float y, float z){
+			min_set = 1;
+
+			VertexMin.x = x;
+			VertexMin.y = y;
+			VertexMin.z = z;
+
+			if(max_set==1){
+				if(VertexMax.x < VertexMin.x || VertexMax.y < VertexMin.y || VertexMax.z < VertexMin.z){
+					
+					fprintf(stderr,"VertexMin is not logical, VertexMin somehow is bigger than VertexMax\n");
+					printf("VertexMin.x:%f VertexMin.y:%f VertexMin.z:%f\n",VertexMin.x,VertexMin.y,VertexMin.z);
+					printf("VertexMax.x:%f VertexMax.y:%f VertexMax.z:%f\n",VertexMax.x,VertexMax.y,VertexMax.z);
+					exit(1);
+				}
+				_is_set=1;
+			}
+		}
+		void setVertexMax(float x, float y, float z){
+			max_set = 1;
+
+			VertexMax.x = x;
+			VertexMax.y = y;
+			VertexMax.z = z;
+
+			if(min_set==1){
+				if(VertexMax.x < VertexMin.x || VertexMax.y < VertexMin.y || VertexMax.z < VertexMin.z){
+					fprintf(stderr,"VertexMax is not logical, VertexMin somehow is bigger than VertexMax\n");
+					exit(1);
+				}
+				_is_set=1;
+			}
+		}
+		Vec3 getVertexMin(){
+			if(_is_set)
+				return VertexMin;
+			fprintf(stderr,"set AABB box first, use copyVertexMin/Max or set VertexMin/Max\n");
+			exit(1);		
+		}
+		Vec3 getVertexMax(){
+			if(_is_set)
+				return VertexMax;
+			fprintf(stderr,"set AABB box first, use copyVertexMin/Max or set VertexMin/Max\n");
+			exit(1);
+		}
+		void copyVertexMin(AABB src){
+			min_set = 1;
+
+			if(src.is_set())
+				VertexMin = src.getVertexMin();
+			else{
+				fprintf(stderr,"the VertexMin/Max of the copied AABB box has not be set\n");
+				exit(1);	
+			}
+			
+			if(max_set==1){
+				if(VertexMax.x < VertexMin.x || VertexMax.y < VertexMin.y || VertexMax.z < VertexMin.z){
+					fprintf(stderr,"VertexMin is not logical, VertexMin somehow is bigger than VertexMax\n");
```

Comment:
Assuming the tree building routine is correct, is there any chance that this branch will be executed?
If not, an assertion is more suitable.